### PR TITLE
Add site_density filter

### PR DESCRIPTION
**Pull request by benjeffery**
Working locally, using this branch to test on GeL.

## Summary by CodeRabbit

- **New Features**
  - Introduced a new `site_density` filter to enhance data analysis capabilities.
  - Added functionality to calculate and display the fraction of sites passing each filter, including the new `site_density` filter.

- **Documentation**
  - Updated configuration examples to reflect the addition of the `site_density` filter.

- **Configuration Changes**
  - Added new parameters for `window_size` and `threshold_sites_per_kbp` to support the `site_density` filter.
  - Commented out certain filters with a note for future fixes.
  - Adjusted Dask configuration parameters to optimize performance.
  - Updated the cluster configuration details, including the `scheduler_address` and `max_threads`.

- **Bug Fixes**
  - Addressed issues with peak RAM usage during the `vcf_to_zarr` conversion steps.

### CODE REVIEW

**Comment by benjeffery** [comment]
@coderabbitai review

**Comment by benjeffery** [comment]
@coderabbitai review

File: steps.py [hunk 1]
```diff
@@ -198,7 +198,10 @@ def post_subset_filters(input, output, wildcards, config, params):  # noqa: A002
...
         filter_config = config["filters"][wildcards.filter]
         for filter_name, filter_kwargs in filter_config.items():
-            if filter_name not in ds.keys():
+            if (
+                f"variant_{filter_name}_mask" not in ds.keys()
+                and filter_name != "site_density"
+            ):
                 mask = getattr(filters, filter_name)(ds, **(filter_kwargs or {}))
                 # Rename to match sgkit convention
                 filter_name = f"variant_{filter_name}_mask"
```
[the remark this answers is not in the document]
Comment (benjeffery) [reply steps.py:207]: From the xarray docs: "Dataset.keys() → a set-like object providing a view on D's keys"

File: steps.py [hunk 2]
```diff
@@ -208,12 +211,52 @@
...
         for filter_name in filter_config:
             filter_name = f"variant_{filter_name}_mask"
-            mask &= ds[filter_name]
-        mask = mask.rename("variant_mask").chunk(chunks).compute()
-        ds.update({"variant_mask": mask})
+            final_mask &= ds[filter_name]
+        final_mask = final_mask.rename("variant_mask").chunk(chunks).compute()
+        ds.update({"variant_mask": final_mask})
         sgkit.save_dataset(
             ds.drop_vars(set(ds.data_vars) - {"variant_mask"}), ds_dir, mode="a"
```
[the remark this answers is not in the document]
Comment (benjeffery) [reply steps.py:261]: The same filter list is used in the step that makes the filters - so this should never happen.

File: steps.py [hunk 3]
```diff
@@ -328,33 +372,42 @@
...
-            f"Site density - "
+            f"Filters passing fraction - "
             f"{wildcards.subset_name}-{wildcards.region_name}-{wildcards.filter}"
         )
         ax.set_xlabel("Position")
-        ax.set_ylabel("Number of sites passing")
+        ax.set_ylabel("Fraction of sites passing")
         box = ax.get_position()
         ax.set_position(
             [box.x0, box.y0 + box.height * 0.15, box.width, box.height * 0.85]
```
[the remark this answers is not in the document]
Comment (benjeffery) [reply steps.py:413]: `site_density` is not referred to in this section unless it is in the filter configuration right?

**Comment by benjeffery** [comment]
This code is currently triggering an intermittent seg fault in the sgkit/dask code. Probably  related to https://github.com/pystatgen/sgkit/issues/869

**Comment by benjeffery** [comment]
I dug in a little, but these segfaults are very tricky in numba code. Using `SGKIT_DISABLE_NUMBA_CACHE` for now.

**Comment by benjeffery** [comment]
This was completely wrong as the density was calculated in site space, not nucleotide space. Rewritten to fix this.

**Comment by benjeffery** [comment]
@coderabbitai review

**Comment by hyanwong** [comment]
> This was completely wrong as the density was calculated in site space, not nucleotide space. Rewritten to fix this.

So easy to do!

File: steps.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (benjeffery) [reply steps.py:207]: There are only ~10

File: steps.py [hunk 4]
```diff
@@ -208,12 +211,80 @@
...
             filter_name = f"variant_{filter_name}_mask"
-            mask &= ds[filter_name]
-        mask = mask.rename("variant_mask").chunk(chunks).compute()
-        ds.update({"variant_mask": mask})
+            final_mask &= ds[filter_name]
+        final_mask = final_mask.rename("variant_mask").chunk(chunks).compute()
+        ds.update({"variant_mask": final_mask})
         sgkit.save_dataset(
             ds.drop_vars(set(ds.data_vars) - {"variant_mask"}), ds_dir, mode="a"
         )
```
[the remark this answers is not in the document]
Comment (benjeffery) [reply steps.py:290]: Gees, come on Mr AI! This uses no for loops and is pretty straightforward, how on earth could it be simpler?

**Comment by benjeffery** [comment]
Added in #39

**Pull request closed without merging**